xds/src/main/java/io/grpc/xds/FaultFilter.java

```diff
@@ -0,0 +1,444 @@
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.protobuf.Any;
+import com.google.protobuf.InvalidProtocolBufferException;
+import com.google.protobuf.TypeRegistry;
+import com.google.protobuf.util.Durations;
+import com.google.protobuf.util.JsonFormat;
+import io.envoyproxy.envoy.extensions.filters.http.fault.v3.HTTPFault;
+import io.envoyproxy.envoy.type.v3.FractionalPercent;
+import io.grpc.CallOptions;
+import io.grpc.Channel;
+import io.grpc.ClientCall;
+import io.grpc.ClientInterceptor;
+import io.grpc.Context;
+import io.grpc.Deadline;
+import io.grpc.LoadBalancer.PickSubchannelArgs;
+import io.grpc.Metadata;
+import io.grpc.MethodDescriptor;
+import io.grpc.Status;
+import io.grpc.internal.DelayedClientCall;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.xds.FaultConfig.FaultAbort;
+import io.grpc.xds.FaultConfig.FaultDelay;
+import io.grpc.xds.Filter.ClientInterceptorBuilder;
+import io.grpc.xds.Matchers.HeaderMatcher;
+import io.grpc.xds.ThreadSafeRandom.ThreadSafeRandomImpl;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nullable;
+
+/** HttpFault filter implementation. */
+final class FaultFilter implements Filter, ClientInterceptorBuilder {
+
+  static final FaultFilter INSTANCE =
+      new FaultFilter(ThreadSafeRandomImpl.instance, new AtomicLong());
+
+  @VisibleForTesting
+  FaultFilter(ThreadSafeRandom random, AtomicLong activeFaultCounter) {
+    this.random = random;
+    this.activeFaultCounter = activeFaultCounter;
+  }
+
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_DELAY_KEY =
+      Metadata.Key.of("x-envoy-fault-delay-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_DELAY_PERCENTAGE_KEY =
+      Metadata.Key.of("x-envoy-fault-delay-request-percentage", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_HTTP_STATUS_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_GRPC_STATUS_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-grpc-request", Metadata.ASCII_STRING_MARSHALLER);
+  @VisibleForTesting
+  static final Metadata.Key<String> HEADER_ABORT_PERCENTAGE_KEY =
+      Metadata.Key.of("x-envoy-fault-abort-request-percentage", Metadata.ASCII_STRING_MARSHALLER);
+  static final String TYPE_URL =
+      "type.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault";
+  private static final String TYPE_URL_TYPED_STRUCT =
+      "type.googleapis.com/udpa.type.v1.TypedStruct";
+
+  private final ThreadSafeRandom random;
+  private final AtomicLong activeFaultCounter;
```

Comment:
Oh, so now since the FaultFilter instance is a singleton, the counter becomes inherently global. Ok.